### PR TITLE
Added PECMedium to IsotropicUniformMediumType, and added PEC2D

### PR DESCRIPTION
**Pull request by caseyflex**

### CODE REVIEW

**Review by tylerflex — APPROVED** [review]
seems ok to me, do we need to add a changelog item in "fixed"?

Asking @shash-sharma for a review too since I'm not totally following the original problem this is fixing.

File: tidy3d/components/grid/mesher.py [hunk 1]
```diff
@@ -373,7 +373,7 @@ def structure_steps(
         min_steps = []
         for structure in structures:
             if isinstance(structure, Structure):
-                if isinstance(structure.medium, (PECMedium, Medium2D)):
+                if structure.medium.is_pec or isinstance(structure.medium, Medium2D):
```
Comment (weiliangjin2021) [inline tidy3d/components/grid/mesher.py:376]: So this use the PEC meshing scheme even if the anisotropic is not PEC along some components?

Comment (caseyflex) [reply tidy3d/components/grid/mesher.py:376]: good catch, I guess it should use the relevant component for meshing. Although, is this a realistic use case? one component PEC, one component not

Comment (weiliangjin2021) [reply tidy3d/components/grid/mesher.py:376]: Maybe for some hyperbolic material that is metallic along some component and dielectric along the rest, and one models the metallic part as PEC.

**Comment by momchil-flex** [comment]
Sorry, what is the status of this? @shash-sharma ?

**Comment by shashwat-sh** [comment]
I think we're waiting on PEC subpixel, and some revisions on this and the associated backend PR

**Comment by momchil-flex** [comment]
PEC subpixel will almost certainly not be in 2.5 - are you saying this and the backend PRs should then also not be labeled as 2.5?

**Comment by caseyflex** [comment]
Weiliang had a proposal on the backend PR that could be implemented before
PEC subpixel as far as I am aware. I haven’t looked into implementing it
yet.

**Comment by caseyflex** [comment]
The issue is wanting a unified treatment of PEC across all three cases:
ordinary, anisotropic, and 2D.

**Comment by caseyflex** [comment]
@weiliangjin2021 do you want to change the meshing to use the relevant component even when another component is PEC?

**Comment by weiliangjin2021** [comment]
> @weiliangjin2021 do you want to change the meshing to use the relevant component even when another component is PEC?

Could you edit `is_comp_pec` method in Medium2D, and push the changes directly? @caseyflex

**Comment by weiliangjin2021** [comment]
One caveat in scene visualization is that only an isotropic PEC will be plotted as PEC. An anisotropic medium where some components are PEC is plotted as regular medium. Do you think it's fine @tylerflex ?

**Comment by caseyflex** [comment]
> > @weiliangjin2021 do you want to change the meshing to use the relevant component even when another component is PEC?
> 
> Could you edit `is_comp_pec` method in Medium2D, and push the changes directly? @caseyflex

It seems like `is_comp_pec` should only be a method in AnisotropicMedium, since it doesn't make sense for an arbitrary AbstractMedium. Then, we can have `is_comp_pec` in Medium2D as well, but in addition to taking the component axis, it can take the normal axis, which is known in the mesher. This would be required to map ss, tt to xx, yy, zz.

**Review by shashwat-sh — APPROVED** [review]
This makes sense to me. About the plotting, I think the current behaviour of only plotting isotropic PEC as PEC makes sense, since the anisotropic case with some PEC components seems to be a fairly special case, and probably used in a very different context.

**Comment by caseyflex** [comment]
@weiliangjin2021 , I pushed the changes we discussed. Also, I changed it to plot as PEC if any component is PEC, rather than if all components are PEC. Please check if this new handling makes sense.

File: tidy3d/components/grid/mesher.py [hunk 2]
```diff
@@ -373,7 +373,13 @@ def structure_steps(
...
         for structure in structures:
             if isinstance(structure, Structure):
-                if isinstance(structure.medium, (PECMedium, Medium2D)):
+                if (
+                    isinstance(structure.medium, AnisotropicMedium)
+                    and structure.medium.is_comp_pec(axis)
+                ) or isinstance(structure.medium, Medium2D):
+                    # for 2d medium, will always ignore even if not PEC;
+                    # later, this will be handled by _grid_corrections_2dmaterials
+                    # in simulation.py
```
Comment (weiliangjin2021) [inline tidy3d/components/grid/mesher.py:382]: Looks a missing case when the medium is PECMedium? (Before this PR, it was `if isinstance(structure.medium, (PECMedium, Medium2D)):`)

**Review by weiliangjin2021 — APPROVED** [review]
One last comment. Otherwise all look good to me.

File: tidy3d/components/grid/mesher.py (hunk 2 above)
Comment (caseyflex) [reply tidy3d/components/grid/mesher.py:382]: good catch, I guess I deleted this by accident.

**Comment by caseyflex** [comment]
> One last comment. Otherwise all look good to me.

thanks for the catch. This should be ready, maybe we want to squash the commits though.

**Comment by weiliangjin2021** [comment]
> > One last comment. Otherwise all look good to me.
> 
> thanks for the catch. This should be ready, maybe we want to squash the commits though.

Thanks, I'll take care of that. Usually we squash just before the merge so that changes can be better monitored.

**Comment by momchil-flex** [comment]
So is there anything left to do @weiliangjin2021 ?

**Comment by weiliangjin2021** [comment]
> So is there anything left to do @weiliangjin2021 ?

just start to squash and merge.

**Pull request merged**